### PR TITLE
[ML][FEATURE] Implement client-side evaluation API

### PR DESCRIPTION
**Pull request by przemekwitek**

### CODE REVIEW

**Comment by przemekwitek** [comment]
run elasticsearch-ci/1 
run elasticsearch-ci/default-distro
run elasticsearch-ci/docbldesx

**Comment by przemekwitek** [comment]
run elasticsearch-ci/1

File: client/rest-high-level/src/main/java/org/elasticsearch/client/ml/EvaluateDataFrameResponse.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+    private static final ParseField SOFT_CLASSIFICATION_METRICS = new ParseField(BinarySoftClassification.NAME);
+
+    @SuppressWarnings("unchecked")
+    private static ConstructingObjectParser<EvaluateDataFrameResponse, Void> PARSER =
+        new ConstructingObjectParser<>(
+            "evaluate_data_frame_response", true, args -> new EvaluateDataFrameResponse((Map<String, EvaluationMetric.Result>) args[0]));
+
+    static {
+        PARSER.declareObject(constructorArg(), (p, c) -> parseSoftClassificationMetrics(p), SOFT_CLASSIFICATION_METRICS);
```
Comment (benwtrent) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/ml/EvaluateDataFrameResponse.java]: Why are we not using `declareNamedObjects` ? You should be able to have more than one named object at the same level and each parsed just fine.

Comment (przemekwitek) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/ml/EvaluateDataFrameResponse.java]: Done.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/ml/dataframe/evaluation/softclassification/AucRocMetric.java [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.constructorArg;
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.optionalConstructorArg;
+
+/**
+ * Area under the curve (AUC) of the receiver operating characteristic (ROC).
+ * The ROC curve is a plot of the TPR (true positive rate) against
+ * the FPR (false positive rate) over a varying threshold.
+ *
+ * This particular implementation is making use of ES aggregations
```
Comment (dimitris-athanasiou) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/ml/dataframe/evaluation/softclassification/AucRocMetric.java]: Some of the commentary here describes implementation details which I think does not need be in the client side.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/ml/dataframe/evaluation/softclassification/AucRocMetric.java [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+        @Override
+        public String toString() {
+            return Strings.toString(this);
+        }
+
+        public static Builder builder() {
+            return new Builder();
+        }
+
+        public static class Builder {
```
Comment (dimitris-athanasiou) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/ml/dataframe/evaluation/softclassification/AucRocMetric.java]: I think the `Builder` shouldn't be necessary

File: client/rest-high-level/src/main/java/org/elasticsearch/client/ml/dataframe/evaluation/softclassification/ConfusionMatrixMetric.java [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        @Override
+        public String toString() {
+            return Strings.toString(this);
+        }
+
+        public static Builder builder() {
+            return new Builder();
+        }
+
+        public static class Builder {
```
Comment (dimitris-athanasiou) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/ml/dataframe/evaluation/softclassification/ConfusionMatrixMetric.java]: We shouldn't need the `Builder` in the client side.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/ml/EvaluateDataFrameResponse.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    private static EvaluationMetric.Result parseMetric(XContentParser parser) throws IOException {
+        String metricName = parser.currentName();
+        try {
+            return parser.namedObject(EvaluationMetric.Result.class, metricName, null);
+        } catch (NamedObjectNotFoundException e) {
+            parser.skipChildren();
+            // Metric name not recognized. Return {@code null} value here and filter it out later.
+            return null;
```
Comment (benwtrent) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/ml/EvaluateDataFrameResponse.java]: It seems to me that we should log something when we find one that is not supported by the client, but IS returned via the server. Having a silent failure inhibits the user from being able to see that their client code is out of step with the server code. 

What say you @dimitris-athanasiou

Comment (przemekwitek) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/ml/EvaluateDataFrameResponse.java]: Skipping unknown metrics makes the parser lenient and allows integrating old client with new server. It is not very different from other places where we drop unknown fields. It's just it's more visible here as I need to do it explicitly.
Happy to hear more from you on logging best practices. From what I see logger is barely used in client code.

Comment (dimitris-athanasiou) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/ml/EvaluateDataFrameResponse.java]: Hm, tricky question. Arguably, we couldn't even log in a higher level than `debug` as if the user is performing an action regularly and consciously, we would flood the log. Given the object parser already silently ignores unknown fields, I would be inclined to treat unknown named fields the same way.

Comment (benwtrent) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/ml/EvaluateDataFrameResponse.java]: OK, sounds good to me! 👍

**Review by benwtrent — APPROVED** [review]

**Review by dimitris-athanasiou — APPROVED** [review]
LGTM

**Comment by przemekwitek** [comment]
run elasticsearch-ci/2

**Comment by przemekwitek** [comment]
run elasticsearch-ci/2

**Comment by przemekwitek** [comment]
run elasticsearch-ci/bwc

**Pull request merged**